samd: Support using a board's external flash chip as boot file system.

robert-hh opened the pull request:
Quite a few boards provide an external on-board flash chip, which this PR supports for the default file system. That provides a larger file system and faster I/O operation, especially for the SAMD51 devices:
- a simple Python flash block device driver, similar to the one already used for the ESP8266. This block device driver is then used with C-coded raw flash drivers.
- a flash driver for the built-in MCU flash, which is the previous samd_flash.c omitting the block device code.
- a QSPI flash driver specific for the SAMD51 series.
- a SPI flash driver, which is port agnostic. It should work with any port providing a SPI and PIN class.
- a changed _boot.py, which enables the appropriate block device/flash device driver combination.

The flash device to be used for the boot file system is defined in the board's mpconfigboard.h.

Note: On some SAMD51 devices, using the QSPI flash for the boot file system increases the code execution speed by about 60%.

dpgeorge commented:
This is definitely a great feature to have.

But I'm curious why you didn't use the existing `drivers/memory/spiflash.[ch]` driver for this?  That driver supports both SPI and QSPI operation and should be generic enough to plug in to the `samd` port.  All that's required is a SPI and/or QSPI backend based on the API defined in `driver/bus/`.

dpgeorge commented on ports/samd/modules/_boot.py:
```diff
@@ -1,19 +1,44 @@
+# Try to mount an external flash drive.
+try:
+    if hasattr(samd, "QSPIflash"):
+        bdev = FlashBdev(samd.QSPIflash())
```
I'm not sure why the `FlashBdev` class is needed, it is a very thin wrapper.  Why not just make `QSPIflash`, `SPIflash` and `MCUflash` implement the block protocol?  Then they can be used directly (and the existing `samd.Flash` could just be renamed to `samd.MCUFlash`).

robert-hh replied:
Thank you for reviewing. `FlashBdevI` is intentionally as small as possible, and I  wanted to be able to support more than one flash type at a device, like external flash and mcuflash. The block device driver was initially combined with the mcuflash driver.

About `drivers/memory/spiflash.[ch]`. I had seen them sometime during the development. The seemed to me pretty specific for Pyboard, and I could not tell how to use them, besides seeming quite large..

dpgeorge commented:
> About drivers/memory/spiflash.[ch]. I had seen them sometime during the development. The seemed to me pretty specific for Pyboard needing more driver code from the sptm32 directory, and I could not tell how to use them, besides seeming quite large..

That SPI flash driver is intended to be used by any port.  It supports both normal SPI and QSPI.  It also supports caching to be able to erase parts that are smaller than the native erase size of the flash, but that's optional via `MICROPY_HW_SPIFLASH_ENABLE_CACHE`.

In the code in this PR there is duplication between the SPI and QSPI driver.  That duplication is gone if you use `drivers/memory/spiflash.c`.  Also the latter code handles errors from timeouts.

IMO it doesn't make sense to have duplicated SPI flash driver code in the samd port and in `drivers/memory`.  They are both doing the same thing.  It would be great if you can try and use the drivers code instead in this PR.  What's needed is:
1. A backend implementation of the (Q)SPI protocol.  There are a few places you can look at to see how it's done: `ports/stm32/qspi.c`, `drivers/bus/softspi.c` and `drivers/bus/softqspi.c`.
2. A frontend for the Python object that implements the block protocol.  In `stm32` this is in `ports/stm32/storage.c` combined with `ports/stm32/spibdev.c`.  But that's pretty complicated and it can be much simpler.

Eventually part (2) above could be made general enough to use by other ports.

robert-hh commented:
Thank you for your hints. Part of these I had seen when considering to use drivers/memory/spiflash.c. 
Bur in the existing PR there is less duplication that it seems, since a board is going to use either SPI of QSPI drivers, and both are made to be as small as possible to fit the constrained flash of especially the SAMD21 boards.  I had the impression, that using a generic approach would result in a substantially larger code size.
Reworking the existing PR according your suggestion seems not reasonable. Instead, it has to be started over again.
With your existing PR, I followed your suggestion include the block device driver part in each and every flash driver, which reduced the code size.

dpgeorge commented:
> Reworking the existing PR according your suggestion seems not reasonable. Instead, it has to be started over again.

Yes I appreciate that.

OK, so let's stick with what we have here for the moment.  Maybe we can refactor at some later stage.

---------------

One thing I would suggest to simplify further what you have here is:
1. hard code the SPI flash parameters in the `SPIflash` constructor, ie always just use the compile time configuration of the SPI pins and baudrate and don't accept any arguments to the `SPIflash` constructor
2. name all the flash types the same, just `samd.Flash()` (only one is enabled at any one time, I assume)
3. this will allow to simplify `_boot.py`, and it could even go back to how it was before this PR

robert-hh commented:
> One thing I would suggest to simplify further what you have here is

Thanks. That sounds good. I'll implement that later today.

robert-hh commented:
> this will allow to simplify _boot.py, and it could even go back to how it was before this PR

One challenge is that the progsize values are different.

dpgeorge commented:
> One challenge is that the progsize values are different.

I suggest using 256 in all cases.

(Or if it really can't spare the RAM for that, either use `gc.mem_free()` to work out the size, or add an ioctl to `samd.Flash()` which returns a suggested optimal value for progsize.)

robert-hh commented:
I just added a `progsize` property to the samd module. The ioctl method looks better, being directly related to the flash.

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,311 @@
+    // 100ms worst case, page write is < 1ms.
+    do {
+        msg[0] = CMD_READ_STATUS;
+        mp_virtual_pin_write(self->cs, 0);
```
I'd suggest using `mp_hal_pin_write(self->cs, 0)`.  It should be smaller (and slightly more efficient) and I don't think you need the generality of the virtual pin API (it's only needed if you want to pass `machine.Signal` objects in).

robert-hh replied:
The intention was to make it port independent. mp_hal_pin_write() is slightly different on the various ports.

dpgeorge replied:
But `mp_hal_pin_write()` should still be portable and work the same way on all ports.  Other drivers (eg dht, soft I2C, soft SPI) use this.

robert-hh replied:
Yes, but the type of the first argument differs between ports. it's mp_hal_pin_obj_t, but that is an mp_obt_t kind on some ports, on others it's an int. Nevertheless, it's easy to change it.

robert-hh replied:
Changed

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,311 @@
+        msg[0] = CMD_READ_STATUS;
+        mp_virtual_pin_write(self->cs, 0);
+        spi_transfer((mp_obj_base_t *)self->spi, 2, msg, msg);
+        mp_virtual_pin_write(self->cs, 1);
```
suggest `mp_hal_pin_write`

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,311 @@
+    uint8_t msg[1];
+
+    msg[0] = CMD_JEDEC_ID;
+    mp_virtual_pin_write(self->cs, 0);
```
`mp_hal_pin_write`

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,311 @@
+    mp_virtual_pin_write(self->cs, 0);
+    spi_transfer(self->spi, 1, msg, NULL);
+    spi_transfer(self->spi, 3, id, id);
+    mp_virtual_pin_write(self->cs, 1);
```
`mp_hal_pin_write`

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,311 @@
+    msg[index++] = (addr >> 16) & 0xff;
+    msg[index++] = (addr >> 8) & 0xff;
+    msg[index++] = addr & 0xff;
+    mp_virtual_pin_write(self->cs, 0);
```
`mp_hal_pin_write`

dpgeorge commented on ports/samd/Makefile:
```diff
@@ -129,6 +131,7 @@ SHARED_SRC_C += \
 	shared/runtime/sys_stdio_mphal.c \
 	shared/timeutils/timeutils.c \
 	shared/tinyusb/mp_cdc_common.c \
+	extmod/virtpin.c \
```
This can be removed if `mp_hal_pin_write` is used (see below).

dpgeorge replied:
Can this be removed now?  (It's anyway included by extmod.mk.)

dpgeorge commented on ports/samd/external_flash_device.h:
```diff
@@ -0,0 +1,461 @@
+/*
```
If we need to have this file, IMO it should go in a common location like `drivers/memory/`.

robert-hh replied:
Sure. Can go there. The file caters for flash chips with different properties.

dpgeorge commented on ports/samd/samd_qspiflash.c:
```diff
@@ -0,0 +1,501 @@
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(samd_qspiflash_ioctl_obj, samd_qspiflash_ioctl);
+
+STATIC const mp_rom_map_elem_t samd_qspiflash_locals_dict_table[] = {
+    { MP_ROM_QSTR(MP_QSTR_flash_size), MP_ROM_PTR(&samd_qspiflash_size_obj) },
```
IMO this method can be removed from all Flash objects.  It's equivalent to ioctl BLOCK_COUNT * BLOCK_SIZE.

robert-hh replied:
That method existed already in the initial version of flash_samd.c and the value is read from the JDEC id. So it's just replicated.  
Note that BLOCK_SIZE is also determined from the JDEC id, with a default of 4096 it that's not possible,

robert-hh replied:
Removed.

robert-hh commented:
b.t.w.: If the code is settled, I would like to re-sort the commits (or suqash them all in one). The structure I suggest:
```
samd/Flash: Rename flash pins consistently.
M   docs/samd/pinout.rst
M   boards/ADAFRUIT_FEATHER_M0_EXPRESS/pins.csv
M   boards/ADAFRUIT_FEATHER_M4_EXPRESS/pins.csv
M   boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/pins.csv
M   boards/MINISAM_M4/pins.csv
M   boards/SEEED_WIO_TERMINAL/pins.csv

samd/Flash: Add the SPI flash driver and set the related board config.
M   boards/ADAFRUIT_ITSYBITSY_M0_EXPRESS/mpconfigboard.h
M   boards/ADAFRUIT_FEATHER_M0_EXPRESS/mpconfigboard.h
M   modmachine.h
M   mphalport.h
M   samd_spiflash.c

samd/Flash: Add the QSPI flash driver  and set the related board config.
M   boards/ADAFRUIT_FEATHER_M4_EXPRESS/mpconfigboard.h
M   boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/mpconfigboard.h
M   boards/MINISAM_M4/mpconfigboard.h
M   boards/SEEED_WIO_TERMINAL/mpconfigboard.h
M   boards/SPARKFUN_SAMD51_THING_PLUS/mpconfigboard.h
M   pin_af.h
M   samd_qspiflash.c
M   external_flash_device.h

samd/Flash: Adapt samd_flash.c, Makefile, mpconfigport.h and _boot.py.
M   Makefile
M   mpconfigport.h
M   modules/_boot.py
M   modsamd.c
M   samd_flash.c

samd/Flash: Extend the code size limit for boards with external flash.
M   boards/ADAFRUIT_FEATHER_M0_EXPRESS/mpconfigboard.mk
M   boards/ADAFRUIT_FEATHER_M4_EXPRESS/mpconfigboard.mk
M   boards/ADAFRUIT_ITSYBITSY_M0_EXPRESS/mpconfigboard.mk
M   boards/ADAFRUIT_ITSYBITSY_M4_EXPRESS/mpconfigboard.mk
M   boards/ADAFRUIT_METRO_M4_EXPRESS/mpconfigboard.mk
M   boards/MINISAM_M4/mpconfigboard.mk
M   boards/SEEED_WIO_TERMINAL/mpconfigboard.mk
M   boards/SPARKFUN_SAMD51_THING_PLUS/mpconfigboard.mk
```

robert-hh commented:
I have updated the PR, but not yet the commits rearranged.

robert-hh committed:
samd/boards: Rename flash pins consistently for QSPI and SPI.

For SAMD21 devices, the board flash signals must be named in pins.csv as
FLASH_MOSI, FLASH_MISO, FLASH_SCK, FLASH_CS for creating the SPI object.

And rename the QSPI pins to QSPI_xxxx instead of FLASH_xxx.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/samd_spiflash: Add SPI flash driver and configure it accordingly.

The SPI flash driver includes the block device for being used as a
filesystem.  It provides the same methods as the driver for the internal
flash.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/samd_qspiflash: Add QSPI flash driver and configure it accordingly.

The QSPI driver provides the interface for using an on-board QSPI flash for
the filesystem.  It provides the same methods as the driver for the
internal flash and uses the same name.  Therefore, only one of the drivers
for internal flash, SPI flash and QSPI flash must be enabled at a time.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd: Adapt existing samd.Flash and integrate with (Q)SPI flash in boot.

Checks are added to ensure, that only one of the flash drivers is selected.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
samd/boards: Extend the code size limit for boards with external flash.

Code size limits are charged to:
- SAMD21:    184K -> 248K
- SAMD51x19: 368K -> 496K
- SAMD51x20: 368K -> 1008K

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
The commits are re-arranged now.

robert-hh commented:
The code is now switched to progsize=256 for all flash types. That is simple and avoids an IOCTL call with an arbitrary chosen function code. If for the SAMD21 RAM is low, then still the stack size could be decreased from 8k.

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,303 @@
+ * Based on spiflash.py by Christopher Arndt and the EEPROM library of Peter Hinch
+ *
+ * https://github.com/SpotlightKid/micropython-stm-lib/tree/master/spiflash
+ * https://github.com/peterhinch/micropython_eeprom.git
```
I see that Peter's repo has an MIT license, which is good.

But for Christopher's, there doesn't seem to be any license at all.  And the spiflash.py there points to another repo which again doesn't have a license.

We need to be extra careful about this because I can see this SPI flash driver being one day moved to a port-agnostic location (eg drivers/ or extmod/) and used by other ports.  Then if there's any ambiguity about the license it will start to affect other parts of the code base.

What here is copied from Christopher's driver?  Can we clarify the license of that driver?

robert-hh replied:
Thank you for spending time on the weekend. Actually, the reference to spiflash.py and the EEPROM driver is obsolete. That not was copied from Peter's driver into spiflash.py, which is a simplified version of Peter's driver and which initially was used for SPI flash. But that's not the case any more, since the driver was rewritten in C and uses no EEPROM code at all. And as far as I can tell the only similarity to Peter's code is the variable name addr4b. But the logic behind that is changed.
So I'll remove the reference.

robert-hh replied:
Looking into the spiflash driver of Christopher, it is part of his micropython-stm-lib, which has a  MIT license file base directory.

dpgeorge replied:
Thanks for checking, clarifying and updating this.

I'm happy with it now.

robert-hh commented:
So I moved the reference to Peter's code in the copyright note, removed the reference to the not used EEPROM parts of Christopher and renamed some variables to more expressive words. Force pushed in the samd_spiflash.c file.

robert-hh commented:
This last change makes the spi_flash object a singleton. It was dynamic before without a need. Still using spi_flash decreases the code size by >368 bytes compared to the driver for internal flash. Using qspi_flash increases the code size by 12 bytes.

dpgeorge commented on ports/samd/samd_qspiflash.c:
```diff
@@ -0,0 +1,492 @@
+    @param iframe iframe register value (configured by caller according to command code)
+    @param addr the address to read or write from. If the instruction doesn't require an address, this parameter is meaningless.
+    @param buffer pointer to the data to be written or stored depending on the type is Read or Write
+    @param size the number of bytes to read or write.
```
This comment style (using `@brief`/`@param`) doesn't match anywhere else in this file or port (?).  And I couldn't find anything similar in the Adafruit code.  So just wondering where it came from, did you write it yourself?

robert-hh replied:
It came form the Adafruit MIT licensed code, which I reused. But I can delete it.

dpgeorge commented on ports/samd/modsamd.c:
```diff
@@ -67,8 +73,25 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_1(samd_pininfo_obj, samd_pininfo);
+    #error Only one type of flash must be used
+    #endif
+    { MP_ROM_QSTR(MP_QSTR_Flash), MP_ROM_PTR(&spiflash_type) },
+    #endif
```
This (messy?) if logic could be simplified by either:
1. having a single name for the flash type, `spiflash_type`, instead of `samd_flash_type` and `samd_qspiflash_type`.  Then the linker will give an error if multiple are defined
2. at the top of this file having a `#define SPIFLASH_TYPE <relevant type>` based on which one is configured.  Then the preprocessor will give an error if multiple are defined.

robert-hh replied:
I went for Option 2. The code looks much more clean.

dpgeorge commented on ports/samd/samd_flash.c:
```diff
@@ -82,9 +76,17 @@ STATIC mp_obj_t samd_flash_init(void) {
+    // No args required. bdev=Flash(). Start Addr & Size defined in samd_flash_obj.
+    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+
+    samd_flash_init();
```
Is it OK if this init is called more than once?  Eg on soft-reset the `_boot.py` will run again and re construct the flash object, calling this function again.

I guess it must be OK because prior to this PR the `flash.init()` method was called each time `_boot.py` ran.

robert-hh replied:
That#s uncritical. samd_flash_init() does nothing but enabling the clock for the device by setting a bit in a register. And that can be done multiple times.

dpgeorge commented on ports/samd/samd_qspiflash.c:
```diff
@@ -124,14 +124,6 @@ static void samd_peripherals_enable_cache(void) {
-    @param addr the address to read or write from. If the instruction doesn't require an address, this parameter is meaningless.
-    @param buffer pointer to the data to be written or stored depending on the type is Read or Write
-    @param size the number of bytes to read or write.
-*/
```
It's OK to leave the comments here, just maybe in a different format that matches existing code, eg:
```c
// Run a single QSPI instruction.
// Parameters are:
// - command: instruction code
// - iframe: ...
```

robert-hh replied:
Done

dpgeorge commented on ports/samd/modsamd.c:
```diff
@@ -35,9 +35,14 @@
 extern const mp_obj_type_t spiflash_type;
+#define SPIFLASH_TYPE spiflash_type
+#endif
+#if MICROPY_HW_MCUFLASH
```
Can this be moved up to line 35, and wrap the `extern const mp_obj_type_t samd_flash_type`?

robert-hh replied:
Sure & done

robert-hh commented:
Updated, and sorry for the confusion.

dpgeorge commented on ports/samd/boards/ADAFRUIT_FEATHER_M0_EXPRESS/pins.csv:
```diff
@@ -8,9 +8,9 @@ PIN_PB03,LED_RX
-PIN_PA08,D4
+PIN_PA08,FLASH_MOSI
+PIN_PA14,FLASH_MISO
+PIN_PA09,FLASH_SCK
```
Should these be renamed?  There's no longer any need to have them accessible in `_boot.py` as these names.

It's a little inconsistent that these are called FLASH and the others below are QSPI.

robert-hh replied:
They are used in samd_spiflash.c. And they are called FLASH_xxxx, to make clear that they are connected to the on-board flash chip, and user accessible SPI pins of a board.

dpgeorge commented on ports/samd/modsamd.c:
```diff
@@ -67,8 +78,8 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_1(samd_pininfo_obj, samd_pininfo);
     { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_samd) },
-    { MP_ROM_QSTR(MP_QSTR_Flash),    MP_ROM_PTR(&samd_flash_type) },
     { MP_ROM_QSTR(MP_QSTR_pininfo),  MP_ROM_PTR(&samd_pininfo_obj) },
+    { MP_ROM_QSTR(MP_QSTR_Flash),    MP_ROM_PTR(&SPIFLASH_TYPE) },
```
This can be put back above in its original order (before pininfo).

robert-hh replied:
Sure.

dpgeorge commented on ports/samd/modules/_boot.py:
```diff
@@ -2,18 +2,17 @@
+    vfs = fs_type(bdev, progsize=256)
 uos.mount(vfs, "/")
 
+del vfs, fs_type, bdev, uos, samd, progsize
```
I'm surprised this doesn't crash... `progsize` isn't defined!

Also, why not delete after the `gc.collect()`, so that `gc` can also be deleted?

dpgeorge commented on ports/samd/samd_flash.c:
```diff
@@ -27,10 +27,13 @@
 #include <stdio.h>
 
 #include "py/runtime.h"
+#include "py/mphal.h"
```
Is this extra include needed?

robert-hh replied:
No.

dpgeorge commented on ports/samd/samd_spiflash.c:
```diff
@@ -0,0 +1,299 @@
+    uint32_t size;
+} spiflash_obj_t;
+
+extern const mp_obj_type_t spiflash_type;
```
I think this should be named `samd_spiflash_type`, to match the other two (flash, qspi).

robert-hh replied:
Done.

robert-hh commented on ports/samd/modmachine.h:
```diff
@@ -39,6 +39,9 @@ extern const mp_obj_type_t machine_dac_type;
 #endif
 extern const mp_obj_type_t machine_pin_type;
+#ifdef MICROPY_HW_QSPIFLASH
+extern const mp_obj_type_t machine_qspiflash_type;
```
Oh yes. That should net be there.

robert-hh commented on ports/samd/samd_flash.c:
```diff
@@ -82,9 +76,17 @@ STATIC mp_obj_t samd_flash_init(void) {
 }
-STATIC MP_DEFINE_CONST_FUN_OBJ_0(samd_flash_init_obj, samd_flash_init);
+
+// FLASH stuff
```
Old, not gold.

robert-hh commented:
Updated again into the last commit. Thank you for the detailed review.

dpgeorge commented:
Thanks for updating.  I broke up and squashed the last commit as appropriate, and tweaked the wording of the commits.

dpgeorge commented:
It's all good.  It's easy for me to fix minor things during merge, and saves going back and forth.

robert-hh commented:
Now that external flash support is enabled, how should the extra 64k flash space used at SAMD21 devices. Options:
- Use LFS2 instead of LFS1: 16k
- Support framebuffer: 4k
- Set Python feature level to FULL_FEATURES instead of CORE_FEATURES: 25k
- Support onewire, dht_readinto with scripts: 3.5 k
- Complex math: 3.2 K
- Viper and Native code: 21k
